modules/glfw/src/filewatcher.cpp

```diff
@@ -0,0 +1,325 @@
+            }
+
+            ++active_;
+        }
+    }
+
+    void watch() {
+        while (!stop_) {
+            {
+                std::scoped_lock lock{mutex_};
+                if (!toRemove_.empty()) {
+                    remove(toRemove_);
+                    toRemove_.clear();
+                }
+                if (!toAdd_.empty()) {
+                    add(toAdd_);
+                    toAdd_.clear();
+                }
+            }
+            if (active_ == 0) {
+                std::this_thread::sleep_for(timeout_);
+            } else {
+                const auto status =
+                    WaitForMultipleObjects(static_cast<DWORD>(active_), handles_.data(), FALSE,
+                                           static_cast<DWORD>(timeout_.count()));
+                if (status >= WAIT_OBJECT_0 && status < WAIT_OBJECT_0 + MAXIMUM_WAIT_OBJECTS) {
+
+                    const auto& path = observed_[status - WAIT_OBJECT_0].first;
+                    auto& files = observed_[status - WAIT_OBJECT_0].second;
+
+                    const auto changedFiles = getChangedAndUpdateFiles(path, files);
+                    for (auto&& [changedFile, action] : changedFiles) {
+                        changeCallback_(path, changedFile, action);
+                    }
+
+                    const auto handle = handles_[status - WAIT_OBJECT_0];
+                    FindNextChangeNotification(handle);
+                }
+            }
+        }
+        for (auto handle : util::as_range(handles_.begin(), handles_.begin() + active_)) {
+            FindCloseChangeNotification(handle);
+        }
+    }
+
+    std::vector<std::pair<std::string, Action>> getChangedAndUpdateFiles(
+        const std::string& path, std::unordered_map<std::string, time_t>& files) {
+
+        std::vector<std::pair<std::string, Action>> changed;
+
+        util::map_erase_remove_if(files, [&](auto& item) {
+            if (!filesystem::fileExists(item.first)) {
+                changed.emplace_back(item.first, Action::Removed);
+                return true;
+            } else {
+                return false;
+            }
+        });
+
+        for (auto& elem : filesystem::getDirectoryContentsRecursively(
+                 path, filesystem::ListMode::FilesAndDirectories)) {
+
+            auto it = files.find(elem);
+            if (it == files.end()) {
+                changed.emplace_back(elem, Action::Removed);
+                files[elem] = filesystem::fileModificationTime(elem);
+            } else {
+                auto newTime = filesystem::fileModificationTime(elem);
+                if (newTime > it->second) {
+                    changed.emplace_back(elem, Action::Modified);
+                    it->second = newTime;
+                }
+            }
+        }
+
+        return changed;
+    }
+
+    static constexpr DWORD filter =
+        FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_LAST_WRITE;
```

Comment:
`FILE_NOTIFY_CHANGE_FILE_NAME` is in there twice. Missing another flag?

Reply:
should be dir not file